### DISCUSSION TITLE
Deleting elements in tensor mechanics

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#16387] [Q&A Modules: Solid mechanics] [answered]
Dear Moose Users,

I was looking at the possibility of deleting elements, when a certain condition, for instance temperature increase, is satisfied.
I have already implemented models where the element stiffness is degraded, however, I would like to compare them
with a model where certain elements are actually removed from the finite element problem.

I have seen something similar is done in XFEM by creating the EFA mesh.
My understanding is that in the XFEM module, a copy of the original mesh is created and manipulated by creating cuts,
then the stress divergence kernel is calculated on the manipulated mesh. But I may be wrong here.

Can someone point me to a strategy to try the element deletion?
Is the creation of a separate mesh necessary?
I would be happy if I could do the deletion just based on element index.

Thank you very much in advance.

Best Regards,
Nicolò Grilli
National University of Singapore

**Comment by MatthewMavis** [comment]
Hi all,
I'm just jumping on this thread to say I'm also interested in this as I'm looking into topology optimisation using moose. I have developed some code to do this based on element IDs though it is currently very rough (https://github.com/MatthewMavis/moose/blob/TopOpt/framework/src/meshgenerators/ElementDeleter.C). 

You should be able to hook into the ElementDeletionGeneratorBase.C like I have which will delete the elements, though one problem I'm currently having with this is that the element ID's are reassigned meaning any data from the simulation cannot be used with the original mesh. So if anyone has any ideas on how to preserve the original IDs it would be a great help.

I hope this helps in some way.

Best Regards,
Matt Mavis

**Reply by ngrilli** [reply]
Dear Matthew,
Thank you very much for your reply.
I will look into the details of your code and give you feedback.
For sure the issue with the elements ID is a difficult one.
Nicolò

**Comment by WilkAndy** [comment]
I think this would be useful to me too, but don't have time to implement it.  I wrote the BlockDeleter, which may give you some hints about what to do, if you choose to do it...

**Reply by ngrilli** [reply]
Dear Andy,
Thank you very much for your reply.
Indeed, I think if block with specific IDs are deleted, the other blocks will maintain the original ID.
So this may be a better solution for me.
I will let you know how it goes
Nicolò

**Comment by ngrilli** [comment]
Dear Moose Users,

I developed the following strategy to delete elements:
1) Generate a new block with new id using ElementSubdomainIDGenerator
2) Deleting that block using BlockDeletionGenerator
3) Of course this will require a restart of the simulation, I restart the simulations at the point where I want to delete the elements using the
SolutionUserObject and SolutionFunction as explained here:
https://mooseframework.inl.gov/modules/porous_flow/restart.html
4) I have tested and seems the variables are assigned correctly at the correct spatial position in the elements remaining after restart
Of course the Element Global ID changes

You can find a simple example here:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/DeleteElementsRestart

Now the question is: will such a method preserve the Material Properties declared in the material blocks?
For instance, I have plastic deformation and I don't want it to be reinitialized to zero after restart.
Can you envisage a method to do so?
I mean restarting the simulation with elements deleted, but preserving the Material Properties at the correct spatial positions in the remaining elements. 

Best Regards,
Nicolò Grilli
National University of Singapore

**Comment by jiangwen84** [comment]
@ngrilli The element deletion method is somehow on our radar to work on in the future. Instead, we have started to implement element activation for AM application recently, see https://github.com/idaholab/moose/pull/16008  It does the opposite thing to deletion, but the method would be similar. It handles material initialization as you asked about in that PR.

**Reply by jiangwen84** [reply]
@dschwen and @dewenyushu they can comment on this better.

**Reply by ngrilli** [reply]
Dear Jiang Wen,

Thank you for your answer, it would be great to have that feature in the future, but I guess it's complicated to implement.

For the moment, I am testing the restart with SolutionUserObject and SolutionFunction.
I am trying to understand if the material properties in the non-deleted elements, like plastic deformation gradient, 
are initialized from the values of the previous solution or are reinitialized by the Material Class (like for a new simulation).

I am sure the plastic deformation gradient is taken from the previous solution with the "basic restart" method, 
when the mesh before and after coincide,
but not sure if that is still valid with SolutionUserObject and SolutionFunction.

I will let you know what I find.

Best Regards,
Nicolò Grilli

**Reply by dewenyushu** [reply]
Our element activation capability should be readily extendable to element deletion. This is because we have "active" and "inactive" subdomains, during element activation, we move elements from "inactive" ->"active" subdomain. Element deletion will be the opposite (i.e., move elements from "active" ->"inactive" subdomain), which seems to be a minor change to #16008 . 

The current PR already takes care of material initialization, updates boundary, does not require any change to the existing physics modules, and is computationally efficient since only the "active" dofs will appear in the system.

**Reply by WilkAndy** [reply]
Thanks @dewenyushu , this is a nice addition to MOOSE.   In the past i have simply set the solid-mechanical stiffness of "deleted" elements to zero, or the thermal conductivity to zero, or the permeability to zero, etc (depending on the physics).  But to be able to actually delete the elements is much nicer.

**Comment by ngrilli** [comment]
Dear Dewen,

Thank you very much for your answer,
I will try your modifications and give you feedback.
Yes, I think element deletion is somehow easier than insertion.

I have tried using simulation restart with SolutionUserObject and SolutionFunction,
but the Material Properties are reinitialized in the elements that are not deleted,
therefore this is not helpful for history dependent materials like plasticity.

It would be great to have your modifications released in MOOSE.

Best Regards,
Nicolò

**Reply by dewenyushu** [reply]
Sounds great! I suppose the solution provided in the PR can help with the material property issue. Let me know how it goes.

**Reply by ngrilli** [reply]
Dear @dewenyushu

I have tested your code modification for element activation/deactivation.
It works well in terms of moving elements from the "inactive block" to the "active block".

However, I see that the stress is still calculated on both active and inactive domains
and the boundary conditions I impose on the active domain are inducing stress and displacement also on the inactive domain.

This happens also in your example:
check_element_addition_by_variable.i

You mentioned in your code only the active DOF will appear,
so perhaps I am not understanding well. 
This is the test I was looking at:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/AddElements
Basically half the domain is active, half is inactive.
I apply displacement on a boundary of the active and I see the inactive domain moving as well.
Therefore the DOF of the inactive seem not be inactive. 

Thank you very much in advance for your answer.

Best Regards,
Nicolò

**Reply by dewenyushu** [reply]
@ngrilli thanks for trying this! 

You will need to make sure your variables (displacements and stresses), kernels (TensorMechanics), and materials are block-restricted, e.g., `block = '1'`.

MOOSE will throw out error about no kernel is defined in the inactive block, so you will need to add the following in your input as well.

```
[Problem]
  kernel_coverage_check = false
[]
```

**Reply by ngrilli** [reply]
Dear @dewenyushu 

Yes, your suggestion worked and your code seems to add elements fine for my application.
I have created some test cases where I see the stress appearing in new elements under load:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/AddElements

I will now try to modify the code to move elements from the active to the inactive block.
I am still studying the part of the code that updates the boundary sets and not sure that has to be changed.
For the moment I will modify only the function
ActivateElementsUserObjectBase::execute()

I will keep you updated about the results.
Thank you very much again for your help.

Best Regards,
Nicolò

**Reply by dewenyushu** [reply]
Sounds great! 

Before looking deeper/trying to change the code, I would suggest trying the following:

- Define your kernels, variables, materials on the block that you want to delete elements from, say `block = '2'`
- Then in the `ActivateElementsCoupled` user object, set `active_subdomain_id = '1'`

It may just work. The element activation code is basically moving elements from one subdomain to another. So in theory it should not prevent us from doing element deletion without much code change. Outside of the user object, we can always define our FEProblem in the subdomain that the elements are moved to or from by using the block restricted capability.

**Comment by ngrilli** [comment, marked as answer]
@dewenyushu @MatthewMavis @jiangwen84 

Dear Dewen,

I have completed and tested an implementation where I can do both element activation and deactivation based on a condition:
https://github.com/ngrilli/c_pfor_am

It is a minor modification of your code.
As you suggested I move elements from the inactive block to the active block and vice versa
based on the condition. 

In this way the plastic deformation is reinitialised properly when elements are reactivated.

The "moving interface" is not yet updated correctly during element elimination.
I think more modifications are needed for this.
However, I am not applying any boundary condition on the "moving interface"
and my understanding is that there is stress free boundary condition at the boundary between the
active block and deactivated block.

I am happy if someone can check and test my implementation
I think it is quite interesting for the mechanics community.

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by dewenyushu** [reply]
Thanks @ngrilli for sharing your branch with us. 

I looked at your example `c_pfor_am/test/tests/ElementAddDelete/ElementAddDeleteDispX.i`. One issue may result in incorrect `moving_interface ` update is that you may have specified the wrong `Mesh/sidesets/normal`. It seems that you have left and right subdomains. Therefore, the initial `moving_interface` should have `normal = '1 0 0'` (+x direction) instead of `normal='0 0 1'`(+z direction). Hope this would fix the issue.

Nevertheless, I am happy that most of the part works for your application!

**Comment by Traiwit** [comment]
@ngrilli  @dewenyushu 

Hi guys,

I am very new to MOOSE, been using it for less than a month, but I found it very powerful.

I've been working on mine stimulation, right now I can do the blockdeletion for a simple model
However, I would like to know if it is possible to add the block/element back to the void space (backfill) without altering the strain on surrounding area (strain-free).

to make it clearer, here's the workflow:
create a solid geometry -> remove a block -> run a equalibirum simulation with gravity activate -> the void is deformed due to stress
-> **backfill the material back into the void without altering the deformation from the previous step**.

Another question I have is: when I use 'BlockDeletionGenerator', the block is gone, the issue would also be that: how can i locate the position of the backfill if I no longer have the deleted block in the system.

I read @ngrilli reply back in Jan, but i am not sure if it will works for my problem, could you please clarify, thanks!.

Thank you

Kind regards,
Trai

below is my script for blockdeletion

[Mesh]
[./gen]
  type = FileMeshGenerator
     file = equlstep_step0.e
     # show_info = true
     use_for_exodus_restart = true
     # exodus_extra_element_integers = 's11_aux s22_aux s33_aux'

  []
[mergedBox]
      type = RenameBlockGenerator
      input = gen
      old_block_id = '2'
      new_block_id = '1'
   []
[ed0]
    type = BlockDeletionGenerator
    input = mergedBox
    block_id = 1
  []
[]

**Reply by WilkAndy** [reply]
You are correct: once you have used `BlockDeletionGenerator`, the block simply does not exist and you cannot add the block's elements back to the simulation again.  You could: (1) create another model without the deletion and read the results from the deleted-model during the initialization; or (2) use a different method like setting the young's modulus of your "deleted" elements to zero, solving, and then resetting the young's modulus to the original value.   However, i suggest you use neither of these methods, and instead try the method described by @ngrilli and @dewenyushu above.

**Reply by ngrilli** [reply]
Dear @Traiwit 

You can find a test case for deleting and reactivating element in my repository:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/ElementAddDelete

When an element is reactivated (moved to the active subdomain), it will not affect affect directly the strain in the other elements.
The nodes that are shared between the active elements and the reactivating element will keep the same displacement
that they had at the previous time step.
The displacement in the newly activated nodes will be initialized based on your displacement ICs (default is 0).
The Material Properties will be reinitialized based on your material objects. 

If the element is reactivated at the beginning of the time step
(flag execute_on = timestep_begin in ActDeactElementsCoupled),
it will affect the strain in the neighbouring elements at the end of the time step
because the element will be considered in the equilibrium equations.

Trust this helps. 

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by dewenyushu** [reply]
@Traiwit I agree with what @ngrilli explained above and I think the element deletion and reactivation should help your problem.

In addition to this, I would like to mention that there is a new implementation by @hugary1995 to unify element activation, deletion, and moving interface, which is based on the element activation code Nicolo and I used, but is more general and flexible in terms of moving elements from one subdomain to another. 

https://github.com/idaholab/moose/pull/17101

**Reply by Traiwit** [reply]
@WilkAndy @ngrilli @dewenyushu 

Thank you for all your help, I attempted all the methods you all suggested, but I found "use a different method like setting the young's modulus of your "deleted" elements to zero, solving, and then resetting the young's modulus to the original value." method is the most applicable to my problem.

However, when I changed the young's modulus value of the 'deleted' block back to original and run the elastic equlibirium simulation, it pushes the deleted block back the the pre-deformed shape which i do not want. is there anyway to work around this?

figure1: pre elastic equlibirium simulation, the green block is the deleted block (applied low young's modulus)
![Screenshot from 2021-03-08 14-49-53](https://user-images.githubusercontent.com/79562439/110272444-5e09b200-801e-11eb-948a-0f6f55d45e3c.png)

figure2: after the simulation, the green block got pushed back
![Screenshot from 2021-03-08 14-50-33](https://user-images.githubusercontent.com/79562439/110272446-5ea24880-801e-11eb-98b0-842696f4dfd0.png)

I tried to use "use_displaced_mesh = true" at every possible location in my input file, but still couldn't solve this problem.

Thank you again

Kind regards,
Traiwit from Australia

**Reply by WilkAndy** [reply]
I think you've got to model what actually happens in real life.   What is the material being "added back"?

**Reply by Traiwit** [reply]
Hi Andy,

Exactly the same material as before it got deleted.
as shown below, i have 3 materials = geo (geology), void, and blackfill.

so for block2 the steps are: elastic bulk model > detetion > backfill

The objective here is to backfill the material into the deformed mesh. 

Regards,
Traiwit

[Materials]
  [./strain]
    type = ComputeFiniteStrain
    # eigenstrain_names = ini_stress
    displacements = 'disp_x disp_y disp_z'
  [../]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]

  [./elasticity_tensor_geo]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 20e9
    poissons_ratio = 0.25
    block = '0 3'
    use_displaced_mesh = true
  [../]
  [./elasticity_tensor_void]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 100000
    poissons_ratio = 0.3
    block = '1'
    use_displaced_mesh = true
  [../]
  [./elasticity_tensor_backfill]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 20e9
    poissons_ratio = 0.3
    block = '2'
    use_displaced_mesh = true
  [../]

  [./density_geo]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 2700 # kg/m3
     block = '0 3'
     use_displaced_mesh = true
  [../]
  [./density_void]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 1.225 # kg/m3
    block = '1'
    use_displaced_mesh = true
  [../]
  [./density_backfill]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 2700 # kg/m3
    block = '2'
    use_displaced_mesh = true
  [../]
[]

**Reply by WilkAndy** [reply]
To me, this sounds like a mining scenario.  In mining, the "backfill" is not identical to the unmined rock.  It is significantly softer, and probably has a much larger Poisson's ratio.   You will have to determine exactly what material properties you need to assign to the backfill.  You will also need to set "strain = 0" in the element as it is backfilled (or measure the stress relative to the strained material).

If it is not a mining scenario and you really are backfilling with identical material, then MOOSE might be doing exactly what you want!

**Reply by Traiwit** [reply]
Hi @WilkAndy,

Thank you for your suggestion, yes, I am making a prototype simulation for mining scenarios, and I am new to the field.

I agree the backfill material should have a much lower young modulus, I set it to 0.001xE_orginal.

I am currently looking for a strain-free (=0) way to backfill this material back into the particular block of the system using MOOSE, do you have any suggestion? Since we cannot directly change the value of strain within the input file.

I found 'Compute Eigenstrain' https://mooseframework.inl.gov/source/materials/ComputeEigenstrain.html, will it work if I set all the 'eigen_base  = 0'. I've also found 'ComputeEigenstrainFromInitialStress' https://mooseframework.inl.gov/source/materials/ComputeEigenstrainFromInitialStress.html which sounds like it might work as well.

Thank you for help, I appreciate it

Kind regards,
Traiwit from Australia

**Reply by WilkAndy** [reply]
Hi @Traiwit ,

So, i think you *might* be able to use controls to ensure the strain-setting activates at the correct time, but it might also be easier to write some C++ yourself.

A few related side notes:
 - You're probably aware, but your situation is likely to be inelastic.  I've created lots of rock-mechanics plasticity classes (various versions of Drucker-Prager, Mohr-Coulomb and weak-plane) that might be useful.
 - You will have to be careful about the roof penetrating the floor
 - Do you work for CSIRO?  If so, i can help and we have our own private app designed for mining situations, and you can use it.

a

**Reply by Traiwit** [reply]
Hi @WilkAndy,

Yes, I will look into that. 

Re: side notes

1. Yes I am fully aware, that will be my next step to work on the inelastic (plastic/non-linear) problems. Is the model you mentioned available online or it is internal to CSIRO? It will be very helpful for my project.
2. Will keep that in mind 
3.  I am not with CSIRO, but I know many people there are using MOOSE for their modelling, my PhD colleague is also working with CSIRO on the Hydrogen storage model. 

Again, thank you for your suggestions.

Kind regards,
Traiwit from Australia

**Reply by Traiwit** [reply]
I managed to work with the eiganstrain to prevent the push-back from introducing backfill material into the void.
but does anyone have any idea on how to obtain the exact eigenstrain tensor in this case? I obtained it via trial & error for the test case.

Kind regards,
Traiwit

#backfilling at block 3
  [./elasticity_tensor_backfill]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 500e6
    poissons_ratio = 0.3
    block = '3'
    use_displaced_mesh = true
  [../]
  [./density_backfill]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 2000 # kg/m3
    block = '3'
    use_displaced_mesh = true
  [../]
  
  [./eigen_backfill]
    type = ComputeEigenstrain
    eigenstrain_name = eigen_true
    eigen_base = '-3e-3 0 0 0 0 0 0 0 -2.5e-3'
    block = '3'
    use_displaced_mesh = true
  [../]
  [./eigen_therest]
    type = ComputeEigenstrain
    eigenstrain_name = eigen_true
    eigen_base = '0'
    block = '0 1 2'
    use_displaced_mesh = true
  [../]

**Reply by jlinBE** [reply]
Hi,

I'm working on the same problem alongside Traiwit, and I *think* all we need to do at this point is: 1. let the pre-backfill simulation reach equilibrium.  2. import the final total strain in a new simulation, as the (negative) eigenstrain.

I am pretty sure this would be possible, but I don't know how to make the Moose syntax work this way.  Does anyone have an example to share?

**Reply by WilkAndy** [reply]
I suggest starting a new discussions thread.  It's likely people have "tuned out" to this discussion, and it's not about deleting elements anymore.

I think what you want is:

1. At the start of each timestep, set the Young's modulus of certain elements to "zero" (as you excavate the mine).  The "certain elements" are defined by a time-dependent Function, or AuxVariable, or whatever.  The "zero" is a small number: it should not be exactly zero, otherwise the displacements won't be uniquely defined.   Different elements will be "set to zero" at each timestep. You've already achieved this (have you looked at https://mooseframework.inl.gov/modules/porous_flow/coal_mining.html, by the way).
2. At the start of each timestep, set the strain in certain elements to zero.  The "certain elements" are defined by another time-dependent Function, etc.   Zeroing the strain should not impact the positions of the nodes in any element.  Its purpose is to ensure the stress in these elements is zero (at the start of the timestep).  This means an alternative would be to introduce a new reference strain, strain_ref, which is the strain in those elements at the start of the timestep, and calculate stress using stress = E * (strain - strain_ref).  You haven't be able to achieve this, since there appears to be nothing in MOOSE that enables either of these ideas.

**Reply by jlinBE** [reply]
Hi Andy, 

Thanks for taking the time to reply to us!  Trai is writing up a new thread.

>1.

This is exactly right!  And in fact we did it while referencing the mech-only version of coal_mining in tensor_mechanics/examples/coal_mining :-)

The key differences in our case is we are solving for steady-state at every step instead of transient, and we're using the base tensormechanics stress/strain classes instead of cosserat.

>2. 

Since we are solving for steady-state, we can't define time-dependent functions.  Other than that, this is exactly what we're trying to do.  Instead of taking the strain at the start of the timestep, we're trying to take it from the steady-state solution of the *previous simulation*.

I think the eigenstrain-related materials (ComputeEigenstrain and ComputeVariableEigenstrain) can probably do what I want, but not sure about the syntax.

**Reply by WilkAndy** [reply]
OK, sounds good.  Instead of doing steady-state, you could do a transient quasi-static simulation, which avoids having to read/write files.

**Reply by ngrilli** [reply]
Dear @Traiwit 

You mention that you want to reduce to zero the stiffness of some of your elements
but it seems in the input file blocks attached to your messages that you do not have any object to do so.

You will need to make a new object that degrades the stiffness based on some criterion,
for instance the value of a coupled variable.

You can see an example code for stiffness degradation in my repository:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeElasticityTensorMelting.C
In my case the coupled variable is the temperature.

Best Regards,
Nicolò Grilli
National University of Singapore
